File: 04_Ginort.py
# PY2
# https://www.hackerrank.com/challenges/ginorts/problem

if __name__ == '__main__':
    myinp  = raw_input()

lcase = list()
ucase = list()

oddig = list()
evdig = list()

for x in myinp :
    if  x >= 'a'and x <= 'z' :
        lcase.append(x)
    elif x >= 'A'and x <= 'Z' :
        ucase.append(x)
    else :
        i = int(x)
        if i % 2 == 1 :
            oddig.append(x)
        else :
            evdig.append(x)

result = sorted(lcase)
result.extend(sorted(ucase))
result.extend(sorted(oddig))
result.extend(sorted(evdig))

print ''.join(result)

File: 04_Ginort.java
/*

Lift from ginorts.py <<###

*/
package ginorts;
import java.util.*;

public class Ginorts {
    public static void main(String[] args) {
        
        ConsoleInput inp = new ConsoleInput('>');
        
        while (inp.hasNext())
        {
            StringBuilder lcase = new StringBuilder();
            StringBuilder ucase = new StringBuilder();
            StringBuilder oddig = new StringBuilder();
            StringBuilder evdig = new StringBuilder();

            String myline = inp.next();
            for (char x : myline.toCharArray())
            {
                if  (x >= 'a' && x <= 'z')
                    lcase.append(x);
                else if (x >= 'A' && x <= 'Z')
                    ucase.append(x);
                else
                {
                    int i = (int)x; i -= (int)'0';

                    if (i % 2 == 1)
                        oddig.append(x);
                    else
                        evdig.append(x);
                }
                
            }

            // Make conversions
            StringBuilder result = new StringBuilder((new Sorted(lcase)).toString());
            result.append(new StringBuilder((new Sorted(ucase)).toString()));
            result.append(new StringBuilder((new Sorted(oddig)).toString()));
            result.append(new StringBuilder((new Sorted(evdig)).toString()));
            
            System.out.println(result);
        }
    }

}
